### ISSUE TITLE
[Bug] Crash On Firebase Storage

### ISSUE THREAD

**Issue by henry-dreamo** [firebase/quickstart-unity#1217]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: 2020.3.16f1
  * Firebase Unity SDK version: 8.6.2
  * Source you installed the SDK: .unitypackage
  * Problematic Firebase Component: Storage
  * Other Firebase Components in use: Auth, Functions, FireStore, Analytics, Crashlytics, Installations
  * Additional SDKs you are using: Facebook, AdMob
  * Platform you are using the Unity editor on: Mac(M1)
  * Platform you are targeting: Android
  * Scripting Runtime: IL2CPP

### [REQUIRED] Please describe the issue here:

I am analyzing the logs by accumulating them in Firebase storage.
I don't know exactly how to reproduce it. Occurs intermittently when sending logs to storage.

Previously I was using StorageReference.PutBytesAsync , and I saw someone answer that PutFileAsync is better than PutFileAsync . So I've changed it now and it's still crashing intermittently.

Attempts to send logs to storage are being made only once at a time. The logic to start the transfer by reading the file starts in the created thread, but the response is always executed in the Unity main thread.
I also saw an article stating that the progress monitor problem in this GitHub issue was supposed to be corrected, but it didn't work.

I create and attach a sample project in the same process as I send logs. However, I couldn't find a way to reproduce the original project or the sample project.

My test Android device crashed once on a weak network. It occurred when I continued the app after turning on the phone for a long time. But I'm still not sure if the crash is the same as this crash.

When using Firebase 8.2.0, the log sender crash came out as follows.

![image](https://user-images.githubusercontent.com/88139855/148889622-e7839d1b-44bb-4c4e-832a-5283a6a3806b.png)
![image](https://user-images.githubusercontent.com/88139855/148889693-27430130-2ff7-41a4-b64d-6c2fe2d746ff.png)

Currently, I have updated to Firebase 8.6.2, and the crashes below are occurring more often than the crashes seen before.
(Physics from the crash stack are not used.)

![image](https://user-images.githubusercontent.com/88139855/148889775-2d852e01-c8e2-4d6a-a0b6-385656de60ff.png)
![image](https://user-images.githubusercontent.com/88139855/148889798-7a7d0c61-77e7-4536-a486-137a269a0e59.png)

#### Steps to reproduce:

There is no clear way to reproduce it.
There are cases where the gzip file was sent as a log several times from a weak network and crashed.

#### Relevant Code:

```
// TODO(you): code here to reproduce the problem
```

#### SampleProject
https://github.com/dream-o/FirebaseCrashSample

I created a sample Firebase project and a sample Unity project and configured it with my current project version, 8.6.2. 
The log sending method is configured in the same way as the one I currently send.

**Comment by paulinon** [comment]
Thanks for reporting this, @henry-dreamo. While I try to replicate this, could you confirm if the issue persists using the latest version of the SDK (8.7.0)?

**Comment by henry-dreamo** [comment]
@paulinon 
okay.
I'll try to apply 8.7.0 to the update in two weeks.

**Comment by henry-dreamo** [comment]
We plan to update by applying 8.7.0 next week.
Have you ever found anything?

**Comment by paulinon** [comment]
Hi @henry-dreamo,

I've tried out your sample project, but I haven't encountered a crash so far. Have you encountered any crashes using the sample project? If so, are there any other steps that have to be considered before expecting a crash?

**Comment by henry-dreamo** [comment]
Hi. @paulinon 

Since I couldn't attach an actual project, I made a sample project by implementing it as close as possible to how it works.
We've never had a crash using the sample project, and we haven't been able to artificially reproduce a crash even in a real project.

And, I want to ask if you have tried pressing the button by building the sample project.
The storage bucket only had the files I uploaded and tested.
You have to press the button freely to write the log and send it to the storage.
If you want, I can send you the built app.

In weak network conditions, when the device screen was turned off while the app was running, 
and then turned on again the next day, luckily, there was an experience that was reproduced. 😢

**Comment by paulinon** [comment]
Hi @henry-dreamo,

I was able to build the app, and I was pressing the button during testing. I didn't notice any crash on my end.

With that, I recommend that you update the SDK so that you'll have the latest features and bug fixes. Let me know if this makes a difference.

**Comment by henry-dreamo** [comment]
Hello! @paulinon 
We plan to update to the latest version (8.7.0) this week.
We will observe if the same crash occurs after the update.

**Comment by helpforsoftware** [comment]
version 8.7.0     using unity 2020.3.26f1
problem still exists
in this line=
 fdb = FirebaseFirestore.DefaultInstance;

**Comment by henry-dreamo** [comment]
Hi. @paulinon

We updated to Firebase version 8.7.0. Also, I updated to Unity 2021.2.8 to see if it was related to the Unity version.
Several callstacks are coming out of Firebase's crashlytics improvements.

![image](https://user-images.githubusercontent.com/88139855/152734608-6533f50b-896f-4e51-a985-fd5842f02453.png)

But in most cases there is less RAM left, and I'm guessing this is a memory leak in Firebase storage. The reason is that there aren't many features that we are using other than Firebase storage.
I wonder if repeated file uploads are also a problem.

In the next update, we will try to disable storage for 50% of users with AB Test, and we will observe if a crash occurs even for users who do not use storage with SetCustomKey for them.

I'm not sure yet that the Firebase Storage SDK is experiencing memory leaks, but may I know if there is a way to upload to storage with a REST API? Or is there a way to avoid using the Unity Firebase SDK?
